sycl/doc/extensions/proposed/SYCL_EXT_ONEAPI_GROUP_STATIC_ARRAY.asciidoc

```diff
@@ -0,0 +1,590 @@
+static constexpr size_t
+key_value_memory_required(sycl::memory_scope scope, sycl::range<dimensions> r);
+----
+
+Table 4. `memory_required` and `key_value_memory_required` member functions of `radix_sorter`.
+|===
+|Member function|Description
+
+|`template<std::size_t ItemsPerWorkItem = 1, std::int32_t dimensions = 1>
+static std::size_t memory_required(sycl::memory_scope scope, sycl::range<dimensions> local_range)`
+|Returns size of temporary memory (in bytes) that is required by the radix
+sorting algorithm defined by the sorter calling by `sort_over_group`.
+`ItemsPerWorkItem` is a parameter for `sycl::span<T, ItemsPerWorkItem>`
+that is an input parameter for `sort_over_group`, where `T` is a first template argument
+for `radix_sorter`. The function can be used for `sort_over_group` without `sycl::span`
+as an input parameter if `ItemsPerWorkItem == 1`.
+If `scope = sycl::memory_scope::work_group`,
+`local_range` is a local range of `sycl::nd_range` that was used to run the kernel;
+if `scope = sycl::memory_scope::sub_group`, `local_range` is a sub-group size.
+If other `scope` values are passed, behavior is unspecified.
+
+|`template<typename U, std::size_t ItemsPerWorkItem, std::int32_t dimensions = 1>
+static constexpr size_t
+key_value_memory_required(sycl::memory_scope scope, sycl::range<dimensions> r);`
+|Returns size of temporary memory (in bytes) that is required by the radix key-value
+sorting algorithm defined by the sorter calling by `sort_over_group`
+with `sycl::span<T, ItemsPerWorkItem>` and `sycl::span<U, ItemsPerWorkItem>`
+as input parameters, where `T` is a first template argument for `radix_sorter`.
+If `scope = sycl::memory_scope::work_group`,
+`local_range` is a local range of `sycl::nd_range` that was used to run the kernel;
+if `scope = sycl::memory_scope::sub_group`, `local_range` is a sub-group size.
+If other `scope` values are passed, behavior is unspecified.
+
+|===
+
+=== SYCL Properties for Interfaces with Static Private Arrays
+
+Consider the `sort_over_group` function (following can be applied for
+`exclusive_scan_over_group` and `inclusive_scan_over_group` as well).
+
+Let `N` is a work-group size, `ItemsPerWorkItem` is a number of elements
+that are processed by one work-item.
+
+Sorting is performed across `N * ItemsPerWorkItem` elements in the group.
+Let `r` is a virtual range for sorting of `N * ItemsPerWorkItem` elements.
+Placing of the result data to `ItemsPerWorkItem` elements per each work-item
+could be done in 2 layouts:
+
+a) Data from the `[r + id * ItemsPerWorkItem; r + (id + 1) * ItemsPerWorkItem)` virtual range
+placed into the private memory under the span for `id`-th work-item.
+
+b) `i * N + id` element of `r` fill the `i`-th element of the private memory
+under the span for `id`-th work-item.
+
+To specify a correct layout for placing of resulting data there are 2 properties
+that satisfy SYCL Properties Extension requirements:
```

Comment:
I think that for sorting algorithms layout doesn't matter because everything goes into one memory and sorts. Only output layout makes sense. For the document we can specify that the input's layout is `close`.

Generally speaking, for scans (for example) it's necessary, because different input's layout will lead to different results.
Will the following call be good enough for users?
```cpp
exclusive_scan_over_group(group, sycl::span{input}, sycl::span{output}, std::plus(), sycl::ext::oneapi::experimental::properties<layout, layout>{});
```
Or it's better to have different types for input and output, e.g. `layout_input`, `layout_output`?

Reply:
Wouldn't input layout matter for stable sorts?

A property list isn't allowed to store the same property more than once, so you'll need separate named properties.  I think `input_layout` and `output_layout` are more readable, but if you want "layout" at the beginning then `layout_in` and `layout_out` look ok (and are shorter).  It might also make sense to have just `layout`, for the common case where the input and output layouts are the same.